File: in_Java/0322 Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] temp = new int[amount+1];
        int cur;
        boolean found;
        for(int i = 1; i < amount+1; i++){
            found = false;
            cur = Integer.MAX_VALUE;
            for(int x : coins){
                if(i % x == 0){
                    cur = i / x < cur? i/x : cur;
                    found = true;
                }
                if(i - x >= 0 && temp[i-x] != -1){
                    found = true;
                    if(temp[i-x]+1 < cur) cur = temp[i-x]+1;
                }
            }
            if(!found) temp[i] = -1;
            else temp[i] = cur;
        }
        //for(int i : temp) System.out.println(i);
        return temp[amount];
    }
}

File: in_Python/0322 Coin Change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [0] * (amount+1)
        for i in range(1, amount+1):
            cur = -1
            for c in coins:
                if i - c == 0:
                    cur = 1
                    break
                elif i - c > 0:
                    if dp[i-c] <= 0:
                        continue
                    if cur == -1:
                        cur = dp[i-c] + 1
                    else:
                        cur = min(cur, dp[i-c]+1)
            dp[i] = cur
        #print(dp)
        return dp[-1]